src/components/Microsoft.DotNet.UpgradeAssistant.Extensions/ExtensionServiceCollection.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions
+{
+    public record ExtensionServiceCollection(IServiceCollection Services, IConfiguration Configuration) : IExtensionServiceCollection
+    {
+        public IExtensionOptionsBuilder<TOption> AddExtensionOption<TOption>(string sectionName)
+            where TOption : class, new()
+        {
+            // Register options first so our custom factory isn't overwritten
+            Services.AddOptions();
+
+            // The default options factory cannot create an instance of ICollection<>
+            Services.AddTransient<IOptionsFactory<ICollection<TOption>>, CollectionOptionsFactory<TOption>>();
+            Services.AddTransient<IOptionsFactory<ICollection<FileOption<TOption>>>, CollectionOptionsFactory<FileOption<TOption>>>();
```

Comment:
In order to map the files, we need to be able to get access to the file provider. Since we don't want to require the use of ifileoptions unless the user wants to, the FileOptions registration is kept to enable the mapping behind the scenes (it's an internal implementation). I thought there was a comment as to why it was there - I'll double check to make sure that's still there